### PR TITLE
New specification of Nameserver11 - #993

### PR DESCRIPTION
**Pull request by tgreenx**
## Purpose

This PR proposes a new document specification for Nameserver11 as mentioned in #993 . The corresponding implementation is provided in another PR ([here](https://github.com/zonemaster/zonemaster-engine/pull/1034)).

## Context

Addresses #993 .

### CODE REVIEW

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 1]
```diff
@@ -27,29 +27,57 @@ case will set DEBUG level on messages for non-responsive name servers.
...
+5. For each name server in *Name Server IP* do:
+
+	1. Send the SOA query **with** OPT record **without** EDNS OPTION-CODE to the name server and collect the response.
+   2. If there is no DNS response, then:
+      1. Send the SOA query **without** OPT record to the name server and 
+         collect the response.
+      2. If there is no DNS response, then output *[NO_RESPONSE]* and 
+         go to next server.
+      3. Else (there is a DNS response), then output 
+         *[BREAKS_ON_EDNS]* and go to next server.
```
Comment (pnax) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:52]: missing outcome and link

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 2]
```diff
@@ -27,29 +27,57 @@ case will set DEBUG level on messages for non-responsive name servers.
...
-		3. The pseudo-section response has an OPT record with version set to 0.
-		4. There is no "OPTION-CODE" present in the response.
-	6. Else output *[NS_ERROR]*.
+				1. The SOA is obtained as response in the ANSWER section.
+				2. If the DNS response has the RCODE "NOERROR".
+				3. The pseudo-section response has an OPT record with version set to 0.
+				4. There is no "OPTION-CODE" present in the response.
+			3. Else output *[NS_ERROR]*.
+	5. Else, if the DNS response meet the following criteria,
+      then output *[EDNS_RESPONSE_WITHOUT_EDNS]* and go to next server.
```
Comment (pnax) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:72]: missing outcome and link

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 3]
```diff
@@ -27,29 +27,57 @@ case will set DEBUG level on messages for non-responsive name servers.
...
+      1. It has the RCODE "NOERROR".
+      2. It has no OPT record.
+   6. Else, if the DNS response meet the following criteria,
+      then output *[EDNS_VERSION_ERROR]* and go to next server.
```
Comment (pnax) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:76]: missing outcome and link

**Review by pnax — COMMENTED** [review]
Nice!
Just some comments/remarks on the form (some links are missing, see comments). And could you review the indentation across the document so that the [raw output](https://raw.githubusercontent.com/zonemaster/zonemaster/3967acc03fb51f3817cf33a4e0c3b572bad5256d/docs/specifications/tests/Nameserver-TP/nameserver11.md) is properly aligned and hopefully the [rendered Markdown](https://github.com/zonemaster/zonemaster/blob/3967acc03fb51f3817cf33a4e0c3b572bad5256d/docs/specifications/tests/Nameserver-TP/nameserver11.md) as well.

**Comment by matsduf** [comment]
We should try to have consistent format of the test case specifications. For that purpose there is a template for the specification, [Template01.md]. Also look at recently created or updated test case specifications such as [DNSSEC18] and [draft Zone09].

In this case the change of the test case is great, and then is should match the template.

There is also a specification of DNS query (#1000) that has been reviewed, but it requires a final approval. With that the specification can be further standardised.

[Template01.md]: https://github.com/zonemaster/zonemaster/blob/develop/docs/internal-documentation/templates/specifications/tests/Template01.md
[DNSSEC18]: https://github.com/zonemaster/zonemaster/blob/develop/docs/specifications/tests/DNSSEC-TP/dnssec18.md
[draft Zone09]: https://github.com/zonemaster/zonemaster/pull/870

**Comment by tgreenx** [comment]
I have updated this specification to the latest specification template format. Please review @matsduf.

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 4]
```diff
@@ -3,91 +3,130 @@
...
 ## Inputs
 
 "Child Zone" - The domain name to be tested.
 
-## Ordered description of steps to be taken to execute the test case
+## Summary
+
+Message                                | Level     | Arguments           | Description of when message tag is outputted
+:--------------------------------------|:----------|---------------------|---------------------------------------------
+NS11_UNKNOWN_OPTION_CODE               | WARNING   |                     | The DNS response contains an unknown EDNS option-code
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:49]:
```
Message                                | Level     | Arguments           | Description of when message tag is outputted
:--------------------------------------|:----------|---------------------|---------------------------------------------
NS11_UNKNOWN_OPTION_CODE               | WARNING   |   ns_ip_list       | The DNS response contains an unknown EDNS option-code
```

**Review by matsduf — CHANGES_REQUESTED** [review]

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 5]
```diff
@@ -3,91 +3,130 @@
...
+
 ## Objective
 
 EDNS is a mechanism to announce capabilities of a DNS implementation,
 and is now basically required by any new functionality in DNS such as
 DNSSEC ([RFC 6891]).
 
-[RFC 6891, section 6.1.2] states that any OPTION-CODE values not understood by a
-responder or requestor MUST be ignored. Unknown OPTION-CODE values must be
+[RFC 6891, section 6.1.2] states that any OPTION-CODE values not understood by
```
[the remark this answers is not in the document]
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 4 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 6]
```diff
@@ -3,91 +3,130 @@
...
-		2. If the DNS response has the RCODE "NOERROR".
-		3. The pseudo-section response has an OPT record with version set to 0.
-		4. There is no "OPTION-CODE" present in the response.
-	6. Else output *[NS_ERROR]*.
- 
+   1. Send the SOA query to the name server and collect the response.
+   2. If there is no DNS response, then go to next name server.
+   3. Else, if the DNS response has the RCODE "FORMERR", then go to next name server.
+   4. Else, if there is an "OPTION-CODE" present in the response, then
+      output *[UNKNOWN_OPTION_CODE]* and go to next name server.
```
[the remark this answers is not in the document]
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 7]
```diff
@@ -3,91 +3,130 @@
...
-	output *[UNKNOWN_OPTION_CODE]*. 
-	5. Else, if the DNS response meet the following four criteria,
-      	then just go to the next name server (no error):
-		1. The SOA is obtained as response in the ANSWER section.
-		2. If the DNS response has the RCODE "NOERROR".
-		3. The pseudo-section response has an OPT record with version set to 0.
-		4. There is no "OPTION-CODE" present in the response.
-	6. Else output *[NS_ERROR]*.
- 
+   1. Send the SOA query to the name server and collect the response.
```
[the remark this answers is not in the document]
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:79]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 8]
```diff
@@ -3,91 +3,130 @@
...
+   3. Else, if the DNS response has the RCODE "FORMERR", then go to next name server.
+   4. Else, if there is an "OPTION-CODE" present in the response, then
+      output *[UNKNOWN_OPTION_CODE]* and go to next name server.
+   5. Else, if the DNS response meet the following four criteria,
+      then go to next name server (no error):
+      1. The SOA is obtained as response in the ANSWER section.
+      2. If the DNS response has the RCODE "NOERROR".
+      3. The pseudo-section response has an OPT record with version set to 0.
+      4. There is no "OPTION-CODE" present in the response.
+   6. Else, go to next name server.
```
[the remark this answers is not in the document]
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md]: Fixed

**Comment by matsduf** [comment]
On the F2F we agreed to make the specification the primary source for the msgid, or rather some database that the specification sucks in, and also is used for the implementation. To take a fist step to that, I have created proposal of updated template, #1083, which you also have reviewed and approved. It is not yet approved, but I suggest that we use that as model for the specifications that we write and update now -- that is what I have done.

My comments are based on that assumption.

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 9]
```diff
@@ -3,91 +3,142 @@
...
 
 ## Inputs
 
 "Child Zone" - The domain name to be tested.
 
-## Ordered description of steps to be taken to execute the test case
+## Summary
+
+Message                                | Level     | Arguments           | Description of when message tag is outputted
+:--------------------------------------|:----------|---------------------|---------------------------------------------
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:48]: Rather
```
Message Tag                   | Level   | Arguments            | Message ID for message tag
:-----------------------------|:--------|:---------------------|:--------------------------------------------
```

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 10]
```diff
@@ -3,91 +3,142 @@
...
+NS11_UNKNOWN_OPTION_CODE               | WARNING   | ns_ip_list          | The DNS response contains an unknown EDNS option-code
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:49]: Include the argument and write a true msgid.

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 11]
```diff
@@ -3,91 +3,142 @@
...
+severity level can be changed in the [Zonemaster-Engine profile]. Also see the
+[Severity Level Definitions] document.
+
+The argument names in the Arguments column lists the arguments used in the
+message. The argument names are defined in the [argument list].
+
+
+## Test procedure
+
+In this section and unless otherwise specified below, the term "[DNS Query]"
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:61]: "DNS Query" --> "EDNS Query"

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 12]
```diff
@@ -3,91 +3,142 @@
...
-1. Create a SOA query for the *Child Zone* with an OPT record with 
-   EDNS OPTION-CODE set to anything other than it is already assigned as in the
-[IANA-DNSSYSTEM-PARAMETERS] and no other EDNS options or flags.
+1. Create the following empty sets:
+   1. Name server IP address ("Unknown Option Code")
 
-2. Obtain the set of name server IP addresses using [Method4] and [Method5]
+2. Create an SOA query for the *Child Zone* with an OPT record with 
+   EDNS OPTION-CODE set to anything other than it is already assigned as in
+   the [IANA-DNSSYSTEM-PARAMETERS] and no other EDNS options or flags.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:72]: Above it is referred to "DNS Query and Response Defaults" but it is not utilized here.
```
2. Create a [EDNS Query] with query type SOA, *Child Zone* as query name and with
EDNS OPTION-CODE set to anything other than it is already assigned as in
the [IANA-DNSSYSTEM-PARAMETERS] and no other EDNS options or flags ("SOA Query").
```

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 13]
```diff
@@ -3,91 +3,142 @@
...
-      	then just go to the next name server (no error):
-		1. The SOA is obtained as response in the ANSWER section.
-		2. If the DNS response has the RCODE "NOERROR".
-		3. The pseudo-section response has an OPT record with version set to 0.
-		4. There is no "OPTION-CODE" present in the response.
-	6. Else output *[NS_ERROR]*.
- 
+4. For each name server in *Name Server IP* do:
+
+   1. Send the SOA query to the name server and collect the response.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:79]: "Send the SOA query to the name" --> "Send *SOA Query* to the name"

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 14]
```diff
@@ -3,91 +3,142 @@
...
+   2. If there is no DNS response, then go to next name server.
+   3. Else, if the DNS response has the RCODE "FORMERR" ([IANA RCODE List]), then go to next name server.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:81]: I suggest that we write "has the [RCODE Name] FormErr" with [RCODE Name] being a link to IANA and use the same form of RCODE name as in the list.

File: docs/specifications/tests/Nameserver-TP/nameserver11.md [hunk 15]
```diff
@@ -3,91 +3,142 @@
...
-[NS_ERROR]: #outcomes
-[RFC 6891, section 6.1.2]: https://tools.ietf.org/html/rfc6891#section-6.1.2
-[RFC 6891]: https://tools.ietf.org/html/rfc6891
-[UNKNOWN_OPTION_CODE]: #outcomes
+## Terminology
+
+No special terminology for this test case.
+
+[NS11_UNKNOWN_OPTION_CODE]:             #summary
+
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver11.md:122]: Please put that in the right place in the list of links.

**Review by matsduf — CHANGES_REQUESTED** [review]

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 9 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:48]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 11 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:61]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 12 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:72]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 13 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:79]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 14 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:81]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 15 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:122]: Fixed

File: docs/specifications/tests/Nameserver-TP/nameserver11.md (hunk 10 above)
Comment (tgreenx) [reply docs/specifications/tests/Nameserver-TP/nameserver11.md:49]: Fixed

**Review by matsduf — APPROVED** [review]

**Comment by matsduf** [comment]
Implementation is done in https://github.com/zonemaster/zonemaster-engine/pull/1034

**Pull request merged**